File: src/main/java/xyz/gnarbot/gnar/music/TrackScheduler.java
package xyz.gnarbot.gnar.music;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import xyz.gnarbot.gnar.Bot;
import xyz.gnarbot.gnar.utils.TrackContext;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class TrackScheduler extends AudioEventAdapter {
    private final MusicManager musicManager;
    private final AudioPlayer player;

    private final Queue<AudioTrack> queue;
    private AudioTrack lastTrack;
    private RepeatOption repeatOption = RepeatOption.NONE;

    /**
     * @param player The audio player this scheduler uses
     */
    public TrackScheduler(MusicManager manager, AudioPlayer player) {
        this.musicManager = manager;
        this.player = player;
        this.queue = new LinkedList<>();
    }

    /**
     * Add the next track to queue or play right away if nothing is in the queue.
     *
     * @param track The track to play or add to queue.
     */
    public void queue(AudioTrack track) {
        if (!player.startTrack(track, true)) {
            queue.offer(track);
        }
    }

    /**
     * Start the next track, stopping the current one if it is playing.
     */
    public void nextTrack() {
        if (queue.isEmpty()) {
            Bot.getPlayers().destroy(musicManager.getGuild());
            return;
        }

        player.startTrack(queue.poll(), false);
    }

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        this.lastTrack = track;

        if (endReason.mayStartNext) {
            switch (repeatOption) {
                case SONG: {
                    AudioTrack newTrack = lastTrack.makeClone();
                    newTrack.setUserData(lastTrack.getUserData());
                    player.startTrack(newTrack, false);
                    break;
                }
                case QUEUE: {
                    AudioTrack newTrack = lastTrack.makeClone();
                    newTrack.setUserData(lastTrack.getUserData());
                    queue.offer(newTrack);
                }
                case NONE:
                    nextTrack();
            }
        }
    }

    @Override
    public void onTrackStuck(AudioPlayer player, AudioTrack track, long thresholdMs) {
        musicManager.getGuild().getTextChannelById(track.getUserData(TrackContext.class).getRequestedChannel())
                .sendMessage("**ERROR:** The track " + track.getInfo().title + " is stuck longer than "
                                + thresholdMs + "ms threshold."
                ).queue();
    }

    @Override
    public void onTrackException(AudioPlayer player, AudioTrack track, FriendlyException exception) {
        musicManager.getGuild().getTextChannelById(track.getUserData(TrackContext.class).getRequestedChannel())
                .sendMessage(
                        "**ERROR:** The track " + track.getInfo().title + " encountered an exception.\n"
                                + "Severity: " + exception.severity + "\n"
                                + "Details: " + exception.getMessage()
                ).queue();
    }

    public RepeatOption getRepeatOption() {
        return repeatOption;
    }

    public void setRepeatOption(RepeatOption repeatOption) {
        this.repeatOption = repeatOption;
    }

    public void shuffle() {
        Collections.shuffle((List<?>) queue);
    }

    public AudioTrack getLastTrack() {
        return lastTrack;
    }

    public Queue<AudioTrack> getQueue() {
        return queue;
    }
}

File: src/main/kotlin/xyz/gnarbot/gnar/music/TrackScheduler.kt
package xyz.gnarbot.gnar.music

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import io.sentry.Sentry
import xyz.gnarbot.gnar.Bot
import xyz.gnarbot.gnar.commands.music.embedTitle
import xyz.gnarbot.gnar.commands.music.embedUri
import xyz.gnarbot.gnar.utils.response.respond
import java.util.*

class TrackScheduler(private val bot: Bot, private val manager: MusicManager, private val player: AudioPlayer) : AudioEventAdapter() {
    val queue: Queue<AudioTrack> = LinkedList()
    var repeatOption = RepeatOption.NONE
    var lastTrack: AudioTrack? = null
        private set

    /**
     * Add the next track to queue or play right away if nothing is in the queue.
     *
     * @param track The track to play or add to queue.
     */
    fun queue(track: AudioTrack) {
        if (!player.startTrack(track, true)) {
            queue.offer(track)
        }
    }

    /**
     * Start the next track, stopping the current one if it is playing.
     */
    fun nextTrack() {
        if (queue.isEmpty()) {
            manager.discordFMTrack?.let {
                it.nextDiscordFMTrack(manager)

                //This basically forces it to poll the next track immediately, they skipped it.
                val track = queue.poll()
                player.startTrack(track, false)

                return
            }

            manager.playerRegistry.executor.execute { manager.playerRegistry.destroy(manager.getGuild()) }
            return
        }

        val track = queue.poll()
        player.startTrack(track, false)

        // fixme DI point
        if (bot.options.ofGuild(manager.getGuild()).music.announce) {
            announceNext(track)
        }
    }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        this.lastTrack = track

        if (endReason.mayStartNext) {
            when (repeatOption) {
                RepeatOption.SONG -> {
                    val newTrack = track.makeClone().also { it.userData = track.userData }
                    player.startTrack(newTrack, false)
                }
                RepeatOption.QUEUE -> {
                    val newTrack = track.makeClone().also { it.userData = track.userData }
                    queue.offer(newTrack)
                    nextTrack()
                }
                RepeatOption.NONE -> nextTrack()
            }
        }
    }

    override fun onTrackStuck(player: AudioPlayer, track: AudioTrack, thresholdMs: Long, stackTrace: Array<out StackTraceElement>) {
        track.getUserData(TrackContext::class.java).requestedChannel.let {
            manager.getGuild()?.getTextChannelById(it)
        }?.respond()?.error(
                "The track ${track.info.embedTitle} is stuck longer than ${thresholdMs}ms threshold."
        )?.queue()

        val exc = buildString {
            append("AudioTrack (id: ${track.info.identifier}) surpassed stuck threshold $thresholdMs\n")
            for (line in stackTrace) {
                append(line.toString())
                append("\n")
            }
        }

        Sentry.capture(exc)
        nextTrack()
    }

    override fun onTrackException(player: AudioPlayer, track: AudioTrack, exception: FriendlyException) {
        if (exception.toString().contains("decoding")) {
            return
        }
        track.getUserData(TrackContext::class.java).requestedChannel.let {
            manager.getGuild()?.getTextChannelById(it)
        }?.respond()?.exception(exception)?.queue()

        Sentry.capture(exception)
    }

    private fun announceNext(track: AudioTrack) {
        manager.currentRequestChannel?.let {
            it.respond().embed("Music Playback") {
                desc {
                    buildString {
                        append("Now playing __**[").append(track.info.embedTitle)
                        append("](").append(track.info.embedUri).append(")**__")

                        track.getUserData(TrackContext::class.java)?.requester?.let{it -> manager.getGuild()?.getMemberById(it)}?.let {
                            append(" requested by ")
                            append(it.asMention)
                        }

                        append(".")
                    }
                }
            }.action().queue()
        }
    }

    fun shuffle() = (queue as MutableList<*>).shuffle()
}
